src/modules/launcher/Plugins/Microsoft.Plugin.Program/Programs/Win32Program.cs

```diff
@@ -885,89 +801,78 @@ private static string GetProgramPathFromRegistrySubKeys(RegistryKey root, string
+                       .Equals((app2.Name?.ToUpperInvariant(), app2.FullPath?.ToUpperInvariant()));
             }
 
-            // Ref : https://stackoverflow.com/questions/2730865/how-do-i-calculate-a-good-hash-code-for-a-list-of-strings
             public int GetHashCode(Win32Program obj)
-            {
-                int namePrime = 13;
-                int executablePrime = 17;
-                int fullPathPrime = 31;
+                => (obj.Name?.ToUpperInvariant(), obj.FullPath?.ToUpperInvariant()).GetHashCode();
+        }
 
-                int result = 1;
+        public static List<Win32Program> DeduplicatePrograms(IEnumerable<Win32Program> programs)
+            => new HashSet<Win32Program>(programs, Win32ProgramEqualityComparer.Default).ToList();
 
-                // Using Ordinal since this is used internally
-                result = (result * namePrime) + obj.Name.ToUpperInvariant().GetHashCode(StringComparison.Ordinal);
-                result = (result * executablePrime) + obj.ExecutableName.ToUpperInvariant().GetHashCode(StringComparison.Ordinal);
-                result = (result * fullPathPrime) + obj.FullPath.ToUpperInvariant().GetHashCode(StringComparison.Ordinal);
+        private static Win32Program GetProgramFromPath(string path)
+        {
+            var extension = Extension(path);
+            if (ExecutableApplicationExtensions.Contains(extension))
+            {
+                return ExeProgram(path);
+            }
 
-                return result;
+            switch (extension)
+            {
+                case ShortcutExtension:
+                    return LnkProgram(path);
+                case ApplicationReferenceExtension:
+                    return CreateWin32Program(path);
+                case InternetShortcutExtension:
+                    return InternetShortcutProgram(path);
+                default:
+                    return null;
             }
         }
 
-        // Deduplication code
-        public static Win32Program[] DeduplicatePrograms(ParallelQuery<Win32Program> programs)
+        private static Win32Program GetRunCommandProgramFromPath(string path)
         {
-            var uniqueExePrograms = programs.Where(x => !(string.IsNullOrEmpty(x.LnkResolvedPath) && ExecutableApplicationExtensions.Contains(Extension(x.FullPath)) && x.AppType != ApplicationType.RunCommand));
-            return new HashSet<Win32Program>(uniqueExePrograms, new RemoveDuplicatesComparer()).ToArray();
+            var program = GetProgramFromPath(path);
+            program.AppType = ApplicationType.RunCommand;
+            return program;
         }
 
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "Keeping the process alive but logging the exception")]
-        public static Win32Program[] All(ProgramPluginSettings settings)
+        public static IList<Win32Program> All(ProgramPluginSettings settings)
```

Comment:
This is build the following way
- Get All Paths (in parallel) inside a hashSet so we only resolve each path once.
- For all Paths, remove the ones which are inside the disabledProgramsList
- Remove Paths we know we don't need to resolve (#965 in old code)
- For each Path (in parallel), resolve based on extension.